### PR TITLE
[atenlib] Integrate symbolic bloom export

### PR DESCRIPTION
**Pull request by titaiwangms**
Add ops

### CODE REVIEW

**Comment by justinchuby** [comment]
I will do another review later today.

**Review by BowenBao — COMMENTED** [review]
Add unittests?

File: onnxscript/function_libs/torch_aten/ops/core.py [hunk 1]
```diff
@@ -138,11 +138,11 @@ def aten_affine_grid_generator_backward(
 
     raise NotImplementedError()
 
-
+@torch_op("aten::alias", trace_only=True)
```
[the remark this answers is not in the document]
Comment (BowenBao) [reply onnxscript/function_libs/torch_aten/ops/core.py:141]: cc @justinchuby

File: onnxscript/function_libs/torch_aten/ops/core.py [hunk 2]
```diff
@@ -4523,6 +4531,13 @@ def aten_segment_reduce(
 
     raise NotImplementedError()
 
+@torch_op("aten::select")
+def aten_select(
```
[the remark this answers is not in the document]
Comment (titaiwangms) [reply onnxscript/function_libs/torch_aten/ops/core.py:4535]: I actually had no idea of writing op here. I am trying to write down necessary ops and wait @justinchuby to the rescue.

File: onnxscript/function_libs/torch_aten/ops/core.py [hunk 3]
```diff
@@ -2935,11 +2942,12 @@ def aten_margin_ranking_loss(
 
     raise NotImplementedError()
 
-
+@torch_op("aten::masked_fill", trace_only=True)
```
[the remark this answers is not in the document]
Comment (titaiwangms) [reply onnxscript/function_libs/torch_aten/ops/core.py:2945]: There is a bug if I tried to make it an onnx function. Looking into this.

File: onnxscript/function_libs/torch_aten/ops/core.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (titaiwangms) [reply onnxscript/function_libs/torch_aten/ops/core.py]: Would love to know more about the mutation with symbolic_fn

[the remark this answers is not in the document]
Comment (justinchuby) [reply onnxscript/function_libs/torch_aten/ops/core.py]: We don’t have mutations in onnx and assumes fx handle them all. Identity sounds like the right thing to do.

File: onnxscript/function_libs/torch_aten/ops/core.py [hunk 4]
```diff
@@ -95,10 +95,13 @@ def aten_addmm(
 ) -> TFloat:
     # addmm(Tensor self, Tensor mat1, Tensor mat2, *, Scalar beta=1, Scalar alpha=1) -> Tensor
 
+    # TODO(titaiwang): op.Gemm seems needed to take care of corner case according to old symbolic_fn.
+    # Currently, it shows op-level validation failing on bloom.
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_aten/ops/core.py:99]: What were the cases? Would replacing the body with Gemm make more sense?

File: onnxscript/function_libs/torch_aten/ops/core.py [hunk 5]
```diff
@@ -526,7 +529,7 @@ def aten_avg_pool1d(
 
     raise NotImplementedError()
 
-
+@torch_op("aten::baddbmm")
 def aten_baddbmm(
     self: TensorType,
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_aten/ops/core.py:534]:
```suggestion
    self: TFloat,
```

Does TFloat seem to be an appropriate type for this? We want a type that is the broadest supported type by the ONNX ops used in the function. You can refer to tensor_typing.py to find or create a type.

File: onnxscript/function_libs/torch_aten/ops/core.py [hunk 6]
```diff
@@ -2935,11 +2945,12 @@ def aten_margin_ranking_loss(
 
     raise NotImplementedError()
 
-
+@torch_op("aten::masked_fill")
 def aten_masked_fill(self: TensorType, mask: TensorType, value: TensorType) -> TensorType:
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_aten/ops/core.py:2949]: I would update the types like above here and for all new functions. LMK if you needed more info.

**Review by justinchuby — COMMENTED** [review]
You may use https://github.com/microsoft/onnx-script/pull/301/files#diff-db423ee0dcd2d4a7c3dddb5141d8b3e2b4cc3f60b32354e36ec14ac3f61dbc01 as an example for enabling tests.

File: onnxscript/function_libs/torch_aten/ops/core.py (hunk 6 above)
Comment (titaiwangms) [reply onnxscript/function_libs/torch_aten/ops/core.py:2949]: I see. I will try to replace annotations

File: onnxscript/function_libs/torch_aten/ops/core.py (hunk 5 above)
Comment (titaiwangms) [reply onnxscript/function_libs/torch_aten/ops/core.py:534]: In aten_mul, why don't we care about uint8?
https://github.com/microsoft/onnx-script/blob/78ea55b888de88bfdadce7c3f6f3f83fa1404c7f/onnxscript/function_libs/torch_aten/ops/core.py#L3440
https://github.com/onnx/onnx/blob/main/docs/Changelog.md#Mul-14

Comment (justinchuby) [reply onnxscript/function_libs/torch_aten/ops/core.py:534]: We probably should. It must be a mistake. Since we are only testing with float inputs rn I expect we miss corner cases here and there. #417

File: onnxscript/function_libs/torch_aten/ops/core.py (hunk 4 above)
Comment (titaiwangms) [reply onnxscript/function_libs/torch_aten/ops/core.py:99]: The idea comes from https://github.com/pytorch/pytorch/pull/46570/files#diff-6616e68665d29ba8ab30d14fc99a8459f974fb58368d71daf7bca89701e557db. But it seems that even though we don't use Gemm, we should still be fine. THe op-level validation is failing with mismatch rn.

File: onnxscript/function_libs/torch_aten/ops/core.py (hunk 5 above)
Comment (titaiwangms) [reply onnxscript/function_libs/torch_aten/ops/core.py:534]: > Does TFloat seem to be an appropriate type for this? We want a type that is the broadest supported type by the ONNX ops used in the function. You can refer to tensor_typing.py to find or create a type.

I think you mean the intersection of supported types across all used ONNX ops in that function?

File: onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py [hunk 7]
```diff
@@ -425,6 +425,11 @@ def _where_input_wrangler(
         matcher=lambda sample: sample.input[0].equal(torch.tensor([])),
         reason="cat does not support zero-dim tensors yet",
     ),
+    # skip(
```
Comment (justinchuby) [inline onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py:428]: You may refer to https://github.com/microsoft/onnx-script/blob/72da830c0e85d8afa1c94f1af8b7c4876090b38b/onnxscript/function_libs/torch_aten/ops/core.py#L2841 to see how we support scalar (tldr check rank and return)

File: onnxscript/function_libs/torch_aten/ops/core.py [hunk 8]
```diff
@@ -1477,11 +1481,14 @@ def aten_cumprod_backward(
...
     raise NotImplementedError()
 
-
-def aten_cumsum(self: TensorType, dim: int, dtype: Optional[int] = None) -> TensorType:
+@torch_op("aten::cumsum")
+def aten_cumsum(self: TensorType, dim: int, dtype: int = -1) -> TensorType:
     # cumsum(Tensor self, int dim, *, ScalarType? dtype=None) -> Tensor
-
-    raise NotImplementedError()
+    if dtype == -1:
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply onnxscript/function_libs/torch_aten/ops/core.py:1492]: I would extract the dtype logic out, mark as trace_only, then wrap only the scalar handling and cumsum part into _aten_cumsum_onnx as a function

File: onnxscript/function_libs/torch_aten/ops/core.py [hunk 9]
```diff
@@ -1487,7 +1487,7 @@ def aten_cumprod_backward(
     raise NotImplementedError()
 
 @torch_op("aten::cumsum", trace_only=True)
-def aten_cumsum(self: TRealUnlessInt16OrInt8, dim: int, dtype: int = -1) -> TRealUnlessInt16OrInt8:
+def aten_cumsum(self: TRealUnlessInt16OrInt8, dim: Union[INT32, INT64], dtype: int = -1) -> TRealUnlessInt16OrInt8:
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_aten/ops/core.py:1490]: Can dim be INT32 (supported by cumsum)? Do we need (a) an explicit cast to INT64, or (b) just assume it’s INt64 and let caller handle type conversion?

File: onnxscript/function_libs/torch_aten/ops/core.py (hunk 5 above)
Comment (justinchuby) [reply onnxscript/function_libs/torch_aten/ops/core.py]: You said it better. That’s the idea. I will put it in #388

File: onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py (hunk 7 above)
Comment (titaiwangms) [reply onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py:428]: Nice

File: onnxscript/function_libs/torch_aten/ops/core.py (hunk 9 above)
Comment (titaiwangms) [reply onnxscript/function_libs/torch_aten/ops/core.py:1490]: Ahh, I get it now. Making a subfunction covering as much as content we could from the origin enables a relative compact function layer still, instead of tracing the whole aten function? Which means the curren solution for dtype if branch would be also like this?

Comment (titaiwangms) [reply onnxscript/function_libs/torch_aten/ops/core.py:1490]: > Can dim be INT32 (supported by cumsum)? Do we need (a) an explicit cast to INT64, or (b) just assume it’s INt64 and let caller handle type conversion?

Get this from https://github.com/onnx/onnx/blob/main/docs/Changelog.md#CumSum-14

T2 accepts both INT32 and INT64 though

Comment (justinchuby) [reply onnxscript/function_libs/torch_aten/ops/core.py:1490]: > Ahh, I get it now. Making a subfunction covering as much as content we could from the origin enables a relative compact function layer still, instead of tracing the whole aten function? Which means the curren solution for dtype if branch would be also like this?

Precisely! What do you think?

Comment (titaiwangms) [reply onnxscript/function_libs/torch_aten/ops/core.py:1490]: I think it's great for a temp solution, as at the end of the day, it's a onnx-script issue to support these pywriting anyway. cc @BowenBao This is the discussion about how to deal with unsupported control flow.

File: onnxscript/function_libs/torch_aten/ops/core.py (hunk 6 above)
Comment (titaiwangms) [reply onnxscript/function_libs/torch_aten/ops/core.py]: Done

File: onnxscript/function_libs/torch_aten/ops/core.py [hunk 10]
```diff
@@ -138,11 +138,11 @@ def aten_affine_grid_generator_backward(
 
     raise NotImplementedError()
 
-
+@torch_op("aten::alias", trace_only=True)
 def aten_alias(self: TensorType) -> TensorType:
```
[the remark this answers is not in the document]
Comment (titaiwangms) [reply onnxscript/function_libs/torch_aten/ops/core.py]: Done

File: onnxscript/function_libs/torch_aten/ops/core.py (hunk 4 above)
Comment (titaiwangms) [reply onnxscript/function_libs/torch_aten/ops/core.py:100]: Track by https://github.com/microsoft/onnx-script/issues/424

File: onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py [hunk 11]
```diff
@@ -361,6 +365,8 @@
     "argmax": core_ops.aten_argmax,
     "argmin": core_ops.aten_argmin,
     "clamp": core_ops.aten_clamp,
+    "cumsum": core_ops.aten_cumsum,
+    "nn.functional.gelu": nn_ops.aten_gelu,
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py:369]: this

**Review by justinchuby — APPROVED** [review]
🎆

**Review by justinchuby — APPROVED** [review]
